walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -26,12 +26,29 @@
     if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
       return Arrays.asList();
     }
-    // Get general available time ranges.
-    Collection<String> attendees = request.getAttendees();
-    List<TimeRange> unavailableTimeWithOverlap = getAttendeesUnavailableTime(events, attendees);
+
+    long meetingDuration = request.getDuration();
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    List<String> allAttendees = new ArrayList<>();
+    allAttendees.addAll(mandatoryAttendees);
+    allAttendees.addAll(optionalAttendees);
+    // If there are time slots that both mandatory and optional attendees are available, return
+    // those time; otherwise, return time that fit just the mandatory attendees.
+    List<TimeRange> unavailableTimeWithOverlap = getAttendeesUnavailableTime(events, allAttendees);
     List<TimeRange> unavailableTime = getNoOverlapTime(unavailableTimeWithOverlap);
-    List<TimeRange> availableTime = getAvailableTime(unavailableTime, request.getDuration());
-    return availableTime;
+    List<TimeRange> availableTime = getAvailableTime(unavailableTime, meetingDuration);
+    if (!availableTime.isEmpty()) {
+      return availableTime;
+    } else {
+      List<TimeRange> mandatoryUnavailableTimeWithOverlap =
+          getAttendeesUnavailableTime(events, mandatoryAttendees);
+      List<TimeRange> mandatoryUnavailableTime =
+          getNoOverlapTime(mandatoryUnavailableTimeWithOverlap);
+      List<TimeRange> mandatoryAvailableTime =
+          getAvailableTime(mandatoryUnavailableTime, meetingDuration);
```

Comment:
Optional: Single-use variables can often be refactored to make code more readable. E.g.
```
List<TimeRange> mandatoryUnavailableTimeWithOverlap =
          getAttendeesUnavailableTime(events, mandatoryAttendees);
List<TimeRange> mandatoryUnavailableTime =
          getNoOverlapTime(mandatoryUnavailableTimeWithOverlap);
List<TimeRange> mandatoryAvailableTime =
          getAvailableTime(mandatoryUnavailableTime, meetingDuration);
return mandatoryAvailableTime;
```
could become:
```
List<TimeRange> mandatoryUnavailableTime =
          getNoOverlapTime(getAttendeesUnavailableTime(events, mandatoryAttendees));
return getAvailableTime(mandatoryUnavailableTime, meetingDuration);
```

Also, do you ever envisage calling getAttendeesUnavailableTime() without immediately calling getNoOverlapTime() on the result? If not, I would be tempted to call getNoOverlapTime() within getAttendeesUnavailableTime() (i.e. return non-overlapped time directly). Your calling code could then look like:

```
List<TimeRange> mandatoryUnavailableTime =
          getAttendeesUnavailableTime(events, mandatoryAttendees);
return getAvailableTime(mandatoryUnavailableTime, meetingDuration);
```
or even:
```
return getAvailableTime(getAttendeesUnavailableTime(events, mandatoryAttendees), meetingDuration);
```

Reply:
Yes, I have multiple single-use variables here by using helper functions. I have now adjusted them to be nested functions.